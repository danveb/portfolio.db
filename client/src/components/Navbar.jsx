import { useEffect } from "react";
import { links } from "../constants/links"; 
import { resumeLink } from "../constants/resume";
import "../styles/Navbar.scss"; 

const Navbar = ({ menuOpen, setMenuOpen }) => {
    // useEffect 
    // when resizing window we'll handle state back to closing mobileMenu/navbar bg
    useEffect(() => {
        const handleResize = () => {
            if(menuOpen) {
                setMenuOpen(!menuOpen); 
            };
        };

        window.addEventListener("resize", handleResize); 

        return () => {
            window.removeEventListener("resize", handleResize); 
        };
    }, [menuOpen, setMenuOpen]);

    return (
        <div className={"navbar " + (menuOpen && "active")}>
            <div className="navbar-left">
                <a href="#intro" className="logo">db<span>.</span></a>
            </div>
            <div className="navbar-right">
                <ul className="navbar-links">
                    {links.map((link) => (
                        <li key={link.id}>
                            <a href={link.url}>{link.text}</a>
                        </li>
                    ))}
                    <li>
                        {resumeLink.map((link) => (
                            <div key={link.id}>
                                <a href={link.url} rel="noopener noreferrer" target="_blank">resume</a>
                            </div>
                        ))}
                    </li>
                </ul>
                <div className="hamburger" onClick={() => setMenuOpen(!menuOpen)}>
                    <span className="line1"></span>
                    <span className="line2"></span>
                    <span className="line3"></span>
                </div>
            </div>
        </div>
    )
}

export default Navbar