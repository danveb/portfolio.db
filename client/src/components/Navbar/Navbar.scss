.navbar {
    width: 100%; 
    height: 50px; 
    display: flex; 
    justify-content: space-between; 
    align-items: center; 
    position: fixed; 
    z-index: 3; 
    background-color: var(--rich-black); 
    color: var(--alice-blue); 
    padding: 0 40px; 
    transition: all 1s ease; 

    @media screen and(max-width: 768px) {
        padding: 0 50px; 
    }

    .navbar-left {
        display: flex; 
        align-items: center; 

        .logo {
            font-size: 30px; 
            font-weight: 600; 
            text-decoration: none; 
            color: inherit; 
            margin-right: 40px; 
        }
    }

    .navbar-right {

        .navbar-links {
            display: flex; 
            align-items: center; 
            justify-content: flex-end; 
            list-style: none; 

            li {
                margin: 0 15px; 
                cursor: pointer; 

                a {
                    color: inherit; 
                    text-decoration: none; 
                    transition: all 0.3s ease; 

                    &:hover {
                        font-weight: 700; 
                        color: var(--han-purple); 
                    }
                }
            }

            @media screen and(max-width: 768px) {
                display: none; 
            }
        }

        .hamburger {
            display: flex; 
            flex-direction: column; 
            justify-content: space-between; 
            width: 25px; 
            height: 20px; 
            cursor: pointer; 

            span {
                width: 100%; 
                height: 3px; 
                background-color: var(--alice-blue); 
                transform-origin: left; 
                transition: all 0.7s ease; 
            }

            @media screen and(min-width: 768.9px) {
                display: none; 
            }
        }
    }

    &.active {
        background-color: var(--prussian-blue); 
        color: var(--alice-blue); 

        .hamburger {

            span {
                &:first-child {
                    background-color: var(--alice-blue); 
                    transform: rotate(45deg); 
                }

                &:nth-child(2) {
                    opacity: 0; // hide color
                }

                &:last-child {
                    background-color: var(--alice-blue); 
                    transform: rotate(-45deg); 
                }
            }
        }
    }
}