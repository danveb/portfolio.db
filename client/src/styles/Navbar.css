header {
  position: sticky; 
  top: 0; 
  left: 0; 
  z-index: 3; 
  transition: all 0.4s ease-out; 
  background-color: var(--rich-black);
}

.navbar {
  height: 50px; 
  width: 100%; 
}

.navbar__wrapper {
  height: 100%; 
  width: 100%; 
  max-width: 768px; 
  margin: 0 auto; 
  display: flex; 
  justify-content: space-between; 
  align-items: center; 
}

.navbar__left .navbar__logo {
  text-decoration: none; 
  color: #fff; 
  font-size: 26px; 
}

.navbar__left span {
  color: var(--bottle-green); 
}

.navbar__mid ul {
  display: flex; 
  gap: 35px; 
  align-items: center; 
}

.navbar__mid li {
  list-style: none; 
}

.navbar__mid a {
  text-decoration: none; 
  color: #fff; 
  background: linear-gradient(0deg, dodgerblue, dodgerblue) no-repeat right bottom / 0 var(--bg-h);
  transition: background-size 350ms;
  --bg-h: 100%; 
  font-size: 14px; 
}

.navbar__mid a:where(:hover, :focus-visible) {
  background-size: 100% var(--bg-h);
  background-position-x: left;
}

.navbar__right {
  display: flex; 
  align-items: center; 
  gap: 15px; 
}

.github-logo {
  width: 25px; 
  height: 25px; 
  cursor: pointer; 
}

.hamburger__menu {
  display: flex; 
  flex-direction: column;
  align-items: center; 
  justify-content: space-between; 
  width: 18px;
  height: 15px; 
  cursor: pointer; 
}

.hamburger__menu span {
  width: 100%; 
  height: 1.5px; 
  background-color: var(--alice-blue); 
  transform-origin: left; 
  transition: all 0.7s ease; 
}

.navbar.active {
  background-color: var(--prussian-blue); 
  color: var(--alice-blue); 
  transition: all 0.7s ease-in-out; 
} 

.navbar.active .hamburger__menu span:first-child {
  transform: rotate(47deg); 
} 

.navbar.active .hamburger__menu span:nth-child(2) {
  opacity: 0; 
}

.navbar.active .hamburger__menu span:last-child {
  transform: rotate(-49deg); 
}

@media screen and (max-width: 1024px) {
  .navbar__wrapper {
    padding: 0 20px; 
  }

  .github-logo {
    width: 21px; 
    height: 21px; 
  }
}

@media screen and (min-width: 768.9px) {
  .hamburger__menu {
    display: none; 
  }
}

@media screen and (max-width: 768px) {
  .navbar__mid ul {
    display: none; 
  }
}