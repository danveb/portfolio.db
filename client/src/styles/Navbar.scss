.navbar {
    width: 100%; 
    height: 50px; 
    position: sticky; 
    top: 0; 
    left: 0; 
    z-index: 3; 
    transition: all 0.4s ease-out; 
    background-color: var(--raisin-black); 

    .navbar__wrapper {
        height: 100%; 
        width: 100%; 
        max-width: 1200px; 
        margin: 0 auto; 
        display: flex; 
        justify-content: space-around; 
        align-items: center; 

        .navbar__left {
            a {
                text-decoration: none; 
                color: #fff; 
                font-size: 30px; 

                span {
                    color: var(--bottle-green); 
                }
            }
        }

        .navbar__mid {
            
            ul {
                display: flex; 
                gap: 35px; 
                align-items: center;
                
                li {
                    list-style: none; 
                    
                    a {
                        text-decoration: none; 
                        color: #fff; 
                        background: linear-gradient(0deg, dodgerblue, dodgerblue) no-repeat right bottom / 0 var(--bg-h);
                        transition: background-size 350ms;
                        --bg-h: 100%;
                    }

                    a:where(:hover, :focus-visible) {
                        background-size: 100% var(--bg-h);
                        background-position-x: left;
                    }
                }

                @media screen and (max-width: 768px) {
                    display: none; 
                }
                
            }
        }

        .navbar__right {
            display: flex; 
            align-items: center; 
            gap: 15px; 


            .github-logo {
                width: 25px; 
                height: 25px; 
                cursor: pointer; 
            }

            .hamburger__menu {
                display: flex;
                flex-direction: column;
                align-items: center; 
                justify-content: space-between;
                width: 18px;
                height: 15px;
                cursor: pointer;

                span {
                    width: 100%; 
                    height: 1.5px; 
                    background-color: var(--alice-blue); 
                    transform-origin: left; 
                    transition: all 0.7s ease; 
                }

                @media screen and (min-width: 768.9px) {
                    display: none; 
                }
            }
        }
    }

    &.active {
        background-color: var(--prussian-blue); 
        color: var(--alice-blue); 
        transition: all 0.7s ease-in-out; 

        .hamburger__menu {

            span {
                
                &:first-child {
                    background-color: var(--alice-blue); 
                    transform: rotate(46deg); 
                }

                &:nth-child(2) {
                    opacity: 0;  
                }

                &:last-child {
                    background-color: var(--alice-blue); 
                    transform: rotate(-50deg); 
                }
            }
        }
    }
}