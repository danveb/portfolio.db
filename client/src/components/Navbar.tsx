import { useEffect } from "react";
import { Link } from "react-router-dom";
import { links } from "../constants/links"; 
import { resumeLink } from "../constants/resume";
import github from "../assets/navbar/github.svg"; 
import "../styles/Navbar.css"; 

import { openProps } from "../types/index"; 

export default function Navbar({ menuOpen, setMenuOpen }: openProps) {
  // useEffect 
  // when resizing window we'll handle state back to closing mobileMenu/navbar bg
  useEffect(() => {
    const handleResize = () => {
      if(menuOpen) {
        setMenuOpen(!menuOpen); 
      };
    };

    window.addEventListener("resize", handleResize); 

    return () => {
      window.removeEventListener("resize", handleResize); 
    };
  }, [menuOpen, setMenuOpen]);

  return (
    <header>
      <nav data-testid="navbarId" className={"navbar " + (menuOpen && "active")}>
        <div className="navbar__wrapper">
          <div className="navbar__left">
            <Link to="/" className="navbar__logo" onClick={() => setMenuOpen(false)}>db<span>.</span></Link>
          </div>
          <div className="navbar__mid">
            <ul>
              {links.map((link) => (
                <li key={link.id}>
                  <Link to={link.url}>{link.text}</Link>
                </li>
              ))}
              {resumeLink.map((link) => (
                <li key={link.id}>
                  <Link to={link.url} rel="noopener noreferrer" target="_blank">resume</Link>
                </li>
              ))}
            </ul>
          </div>
          <div className="navbar__right">
            <Link to={"https://github.com/danveb"} rel="noopener noreferrer" target="_blank" onClick={() => setMenuOpen(false)}>
              <img className="github-logo" src={github} alt="github logo" />
            </Link>
            <div data-testid="hamburger" className="hamburger__menu" onClick={() => setMenuOpen(!menuOpen)}>
              <span></span>
              <span></span>
              <span></span>
            </div>
          </div>
        </div>
      </nav>
    </header>
  )
}
