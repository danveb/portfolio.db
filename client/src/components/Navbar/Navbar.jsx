import { links } from "../../constants/links"; 
import "./Navbar.scss"; 

const Navbar = ({ menuOpen, setMenuOpen }) => {
    return (
        <div className={"navbar " + (menuOpen && "active")}>
            <div className="navbar-left">
                <a href="#intro" className="logo">db<span>.</span></a>
            </div>
            <div className="navbar-right">
                <ul className="navbar-links">
                    {links.map((link) => (
                        <li key={link.id}>
                            <a href={link.url}>{link.text}</a>
                        </li>
                    ))}
                    <li>
                        <a href="https://drive.google.com/file/d/12SyFoMp7IEL_Vkne9W3Akz5Nt0gB9YvE/view?usp=sharing" rel="noopener noreferrer" target="_blank">resume</a>
                    </li>
                </ul>
                <div className="hamburger" onClick={() => setMenuOpen(!menuOpen)}>
                    <span className="line1"></span>
                    <span className="line2"></span>
                    <span className="line3"></span>
                </div>
            </div>
        </div>
    )
}

export default Navbar