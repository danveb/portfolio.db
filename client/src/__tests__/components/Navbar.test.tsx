import { render, screen, fireEvent } from "@testing-library/react"; 
import Navbar from "../../components/Navbar"; 
import { BrowserRouter } from "react-router-dom";

// mockProps
const mockProps = {
  menuOpen: false, 
  setMenuOpen: jest.fn(), 
}; 

describe("Navbar component", () => {
  // smoke test
  test("renders without crashing", () => {
    render(
      <BrowserRouter>
        <Navbar {...mockProps} />
      </BrowserRouter>
    ); 
  }); 

  // snapshot test
  test("matches snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Navbar {...mockProps} />
      </BrowserRouter>
    ); 
    expect(asFragment()).toMatchSnapshot(); 
  });

  // getByText
  test("should display the anchor tag", () => {
    render(
      <BrowserRouter>
        <Navbar {...mockProps} />
      </BrowserRouter>
    ); 
    const anchor = screen.getByText("resume"); 
    expect(anchor).toBeInTheDocument(); 
    expect(anchor).not.toHaveClass("resume"); 
  });

  // getByTestId 
  test("renders with testId", () => {
    render(
      <BrowserRouter>
        <Navbar {...mockProps} />
      </BrowserRouter>
    ); 
    const testId = screen.getByTestId("navbarId"); 
    expect(testId).toBeInTheDocument(); 
  }); 

  // fireEvent on hamburger menu
  test("click event on hamburger menu", () => {
    render(
      <BrowserRouter>
        <Navbar {...mockProps} /> 
      </BrowserRouter>
    );
    const hamburgerMenu = screen.getByTestId("hamburger"); 
    fireEvent.click(hamburgerMenu);  
    expect(mockProps.setMenuOpen).toHaveBeenCalledTimes(1); 
  }); 
});
