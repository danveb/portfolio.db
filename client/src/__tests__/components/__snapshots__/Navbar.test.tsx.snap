// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component matches snapshot 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="navbar false"
      data-testid="navbarId"
    >
      <div
        class="navbar__wrapper"
      >
        <div
          class="navbar__left"
        >
          <a
            class="navbar__logo"
            href="/"
          >
            db
            <span>
              .
            </span>
          </a>
        </div>
        <div
          class="navbar__mid"
        >
          <ul>
            <li>
              <a
                href="/about"
              >
                about
              </a>
            </li>
            <li>
              <a
                href="/projects"
              >
                projects
              </a>
            </li>
            <li>
              <a
                href="/contact"
              >
                contact
              </a>
            </li>
            <li>
              <a
                href="https://drive.google.com/file/d/1N0VvKMKcAfogYAo-3gB3YONRT540EPLr/view?usp=sharing"
                rel="noopener noreferrer"
                target="_blank"
              >
                resume
              </a>
            </li>
          </ul>
        </div>
        <div
          class="navbar__right"
        >
          <a
            href="https://github.com/danveb"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="github logo"
              class="github-logo"
              src="github.svg"
            />
          </a>
          <div
            class="hamburger__menu"
            data-testid="hamburger"
          >
            <span />
            <span />
            <span />
          </div>
        </div>
      </div>
    </nav>
  </header>
</DocumentFragment>
`;
